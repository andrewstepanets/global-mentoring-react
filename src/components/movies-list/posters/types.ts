export interface IMovieDetails {
  id: number;
  budget: number;
  genres: string[];
  overview: string;
  poster_path: string;
  release_date: string;
  revenue: number;
  runtime: number;
  tagline: string;
  title: string;
  vote_average: number;
  vote_count: number;
}

export interface PostersProps {
  setLoadingMovieDetails: boolean;
  setErrorMovieDetails: boolean;
  setMovieDetails: IMovieDetails;
}

export interface PosterItemProps extends PostersProps {
  poster: IMovieDetails;
  genre: JSX.Element[] | string[];
}
