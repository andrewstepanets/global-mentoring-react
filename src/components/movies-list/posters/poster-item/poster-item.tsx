import useMoreDetailsMovie from 'hooks/useMoreDetailsMovie';
import React, { FC, useCallback, useEffect } from 'react';
import shortid from 'shortid';
import defaultImgMovie from '../../../../assets/images/fallback_movie.png';
import {
  PostersGenre,
  PostersImg,
  PostersItem,
  PostersLink,
  PostersTitle,
  PostersTitleYear,
  PostersWrapTitle,
} from './styles';

interface PosterItemProps {
  setMovieDetails: {
    title: string;
    tagline: string;
    vote_average: number;
    vote_count: number;
    release_date: string;
    poster_path: string;
    overview: string;
    budget: number;
    revenue: number;
    runtime: number;
    genres: string[];
    id: number;
  };
  poster: any;
  genre: any;
}

export const PosterItem: FC<PosterItemProps> = ({
  setMovieDetails,
  genre,
  poster,
}: any) => {
  const {
    movieDetails,
    loadingMovieDetails,
    errorMovieDetails,
    fetchMovieDetails,
  }: any = useMoreDetailsMovie();

  const handleMoreDetails = useCallback((id) => {
    fetchMovieDetails(id);
  }, []);

  useEffect(() => {
    setMovieDetails(movieDetails);
  }, [movieDetails]);

  const addDefaultSrc = useCallback(({ target }) => {
    target.src = defaultImgMovie;
    target.alt = 'image not found';
  }, []);

  return (
    <PostersItem key={shortid.generate()}>
      <PostersLink onClick={() => handleMoreDetails(poster.id)}>
        <PostersImg
          src={poster.poster_path}
          alt={poster.title}
          onError={addDefaultSrc}
        />
        <PostersWrapTitle>
          <PostersTitle>{poster.title}</PostersTitle>
          <PostersTitleYear>{poster.release_date}</PostersTitleYear>
        </PostersWrapTitle>
        <PostersGenre>{genre}</PostersGenre>
      </PostersLink>
    </PostersItem>
  );
};
