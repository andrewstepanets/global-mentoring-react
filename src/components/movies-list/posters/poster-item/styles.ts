import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { colors, colorsRgba, rgbaMap } from '../../../../styles/constants';

export const PostersItem = styled.li`
  margin: 15px;
  padding: 15px;
  height: 620px;
  box-shadow: 0 4px 8px 0 ${rgbaMap(colorsRgba.rgba_white, 0.2)};
  transition: 0.3s;
  :hover {
    box-shadow: 0 8px 16px 0 ${rgbaMap(colorsRgba.rgba_white, 0.2)};
  }
`;

export const PostersLink = styled(Link)`
  cursor: pointer;
`;

export const PostersWrapTitle = styled.div`
  display: flex;
  justify-content: space-between;
  width: 300px;
  padding: 10px 0;
  color: ${colors.grey_native};
`;

export const PostersTitle = styled.h2`
  font-size: 20px;
`;

export const PostersTitleYear = styled.span`
  align-self: center;
  padding: 5px;
  font-size: 12px;
  white-space: nowrap;
  border: 1px solid grey;
  border-radius: 4px;
`;

export const PostersImg = styled.img`
  width: 300px;
  height: 455px;
`;

export const PostersGenre = styled.p`
  padding-bottom: 15px;
  width: 300px;
  color: ${colors.grey_native};
`;
