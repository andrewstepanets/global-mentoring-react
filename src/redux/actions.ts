import {
  FETCH_MOVIES,
  DELETE_MOVIE,
  POSTER_ID,
  ADD_MOVIE,
  EDIT_MOVIE,
  FILTER_MOVIES,
  FILTER_ITEM,
  SEARCH_MOVIES,
  ERROR_MESSAGE,
} from './types';

export const fetchMovies = (payload) => {
  return { type: FETCH_MOVIES, payload };
};

export const deleteMovie = (payload) => {
  return { type: DELETE_MOVIE, payload };
};

export const putPosterId = (payload) => {
  return { type: POSTER_ID, payload };
};

export const addMovie = (payload) => {
  return { type: ADD_MOVIE, payload };
};

export const editMovie = (payload) => {
  return { type: EDIT_MOVIE, payload };
};

export const filterMovies = (payload) => {
  return { type: FILTER_MOVIES, payload };
};

export const filterItem = (payload) => {
  return { type: FILTER_ITEM, payload };
};

export const searchMovies = (payload) => {
  return { type: SEARCH_MOVIES, payload };
};

export const getErrorMessage = (payload) => {
  return { type: ERROR_MESSAGE, payload };
};
