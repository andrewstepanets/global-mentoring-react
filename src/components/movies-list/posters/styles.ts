import styled from 'styled-components';
import { colors, mediaQueries } from '../../../styles/constants';

export const PostersWrapper = styled.div`
  padding: 0 30px 30px 30px;
`;

export const NumberMovies = styled.h3`
  padding: 25px 70px;
  font-size: 18px;
  font-weight: 100;
  color: ${colors.white};
  ${mediaQueries('sizeXXL')`
    text-align: center;
  `}
  span {
    font-size: 22px;
    font-weight: 600;
  }
`;

export const PostersList = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  align-items: center;
  margin-bottom: 70px;
`;

export const PostersError = styled.div`
  margin: 150px;
  color: ${colors.white};
  font-weight: 200;
  font-size: 32px;
  text-align: center;
`;
