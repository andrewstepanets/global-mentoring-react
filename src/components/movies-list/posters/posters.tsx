import { API_BASE, API_FILTER, API_PAGE, API_SEARCH } from '@constants';
import { Button } from 'components/button';
import { useApiRequest } from 'hooks/useApiRequest';
import { ParamTypes } from 'pages/types';
import React, { FC, memo, useCallback, useEffect } from 'react';
import Loader from 'react-loader-spinner';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { fetchMovies, filterMovies } from 'redux/actions';
import shortid from 'shortid';
import { encodeURL } from 'utils/utils';
import { PosterItem } from './poster-item';
import {
  NumberMovies,
  PostersError,
  PostersList,
  PostersWrapper,
} from './styles';
import { PostersProps } from './types';

const Posters: FC<PostersProps> = ({
  setMovieDetails,
  setLoadingMovieDetails,
  setErrorMovieDetails,
  hideEdit,
  hideDelete,
}) => {
  const filterItem = useSelector(({ movies: { filterItem } }) => filterItem);
  const movies = useSelector(({ movies: { items } }) => items);
  const currentPage = useSelector(({ movies: { currentPage } }) => currentPage);
  const error = useSelector(({ movies: { error } }) => error);
  const loading = useSelector(({ movies: { loading } }) => loading);
  const totalPages = useSelector(({ movies: { totalPages } }) => totalPages);
  const { fetchData: getMovies } = useApiRequest('get', API_BASE, fetchMovies);
  const { fetchData: getMoreMovies } = useApiRequest(
    'get',
    API_PAGE,
    fetchMovies,
  );
  const { slug } = useParams<ParamTypes>();
  const encode = encodeURL(slug);
  const { fetchData: filteredMovies } = useApiRequest(
    'get',
    encode !== 'undefined'
      ? `${API_SEARCH}${encode}&searchBy=title`
      : `${API_FILTER}${filterItem}`,
    filterMovies,
  );

  useEffect(() => {
    getMovies();
  }, []);

  const handleLoadMoreMovies = useCallback(() => {
    if (filterItem !== 'all' || encode !== 'undefined') {
      filteredMovies(`&offset=${currentPage}`);
    } else {
      getMoreMovies(currentPage);
    }
  }, [currentPage, filterItem]);

  const posters = movies.map((poster) => {
    if (!poster.genres) poster.genres = [];

    const genre = poster.genres.map((genre) => (
      <span key={shortid.generate()}> {genre}, </span>
    ));

    return (
      <PosterItem
        key={shortid.generate()}
        setMovieDetails={setMovieDetails}
        setLoadingMovieDetails={setLoadingMovieDetails}
        setErrorMovieDetails={setErrorMovieDetails}
        poster={poster}
        genre={genre}
        hideEdit={hideEdit}
        hideDelete={hideDelete}
      />
    );
  });

  return (
    <PostersWrapper>
      {error && <PostersError>{error}</PostersError>}
      {loading && (
        <Loader type="Circles" color="#00BFFF" height={80} width={80} />
      )}
      {!error && !loading && (
        <>
          {' '}
          <NumberMovies>
            <span>{movies.length}</span> movie found
          </NumberMovies>
          <PostersList>{posters}</PostersList>
          {totalPages > 1 && (
            <Button
              load
              text="Load more posters"
              type="button"
              onClick={handleLoadMoreMovies}
            />
          )}
        </>
      )}
    </PostersWrapper>
  );
};

export default memo(Posters);
