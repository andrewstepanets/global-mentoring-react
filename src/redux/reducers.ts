import { combineReducers } from 'redux';
import {
  FETCH_MOVIES,
  POSTER_ID,
  DELETE_MOVIE,
  ADD_MOVIE,
  EDIT_MOVIE,
  FILTER_MOVIES,
  FILTER_ITEM,
  SEARCH_MOVIES,
  ERROR_MESSAGE,
} from './types';

export const initialState = {
  items: [],
  currentPage: 1,
  totalPages: 0,
  error: null,
  loading: true,
  posterId: null,
  filterItem: 'all',
};

export const movies = (state = initialState, { type, payload }) => {
  switch (type) {
    case FETCH_MOVIES:
      return {
        ...state,
        items:
          state.currentPage === 1
            ? payload.data
            : [...state.items, ...payload.data],
        currentPage: payload.offset + 1,
        totalPages: payload.totalAmount - state.currentPage,
        loading: false,
      };

    case POSTER_ID:
      return {
        ...state,
        posterId: payload,
      };

    case ADD_MOVIE:
      return {
        ...state,
        items: [{ id: state.posterId, ...payload }, ...state.items],
        loading: false,
      };

    case EDIT_MOVIE:
      const updatedMovie = state.items.find((movie) => {
        return movie.id === payload.id;
      });

      if (updatedMovie) {
        Object.assign(updatedMovie, payload);
      }

      return {
        items: [...state.items],
      };

    case DELETE_MOVIE:
      return {
        ...state,
        items: [...state.items.filter((item) => item.id !== state.posterId)],
        loading: false,
      };

    case FILTER_MOVIES:
      return {
        ...state,
        items: [...state.items, ...payload.data],
        currentPage: payload.offset + 1,
        totalPages: payload.totalAmount - state.currentPage,
        loading: false,
      };

    case FILTER_ITEM:
      return {
        ...state,
        currentPage: 0,
        filterItem: payload,
        items: [],
        loading: true,
      };

    case SEARCH_MOVIES:
      console.log('payload: ', payload);

      return {
        ...state,
        items: payload.data,
        currentPage: 1,
        totalPages: payload.totalAmount - state.currentPage,
      };

    case ERROR_MESSAGE:
      return {
        ...state,
        items: [],
        currentPage: 1,
        totalPages: payload.totalAmount - state.currentPage,
        error: payload,
        loading: false,
      };

    default:
      return state;
  }
};

export default combineReducers({
  movies,
});
