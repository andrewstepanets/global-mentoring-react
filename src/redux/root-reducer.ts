import { combineReducers } from 'redux';
import modals, { ModalsState } from './modals/modals.reducer';
import movies, { MoviesState } from './movies/movies.reducer';

export interface AppState {
  movies: MoviesState;
  modals: ModalsState;
}

const rootReducer = combineReducers({
  movies,
  modals,
});

export default rootReducer;
