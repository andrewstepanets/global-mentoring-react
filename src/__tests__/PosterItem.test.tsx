import * as React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { getByText, render } from '@testing-library/react';

import { PosterItem } from '../components/movies-list/posters/poster-item';
import ModalsContextProvider from '../modal-context/modal-context';

import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { movies } from '../redux/reducers';

const poster = {
  title: 'Fifty Shades Freed',
  tagline: "Don't miss the life",
  vote_average: 6,
  vote_count: 1195,
  release_date: '2016-02-11',
  poster_path:
    'https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg',
  overview: 'Lorem ipsum dolor, sit amet consectetur adipisicing elit.',
  budget: 55000000,
  revenue: 1023784195,
  runtime: 110,
  genres: ['Animation', 'Adventure', 'Family', 'Comedy'],
  id: 337167,
};

const renderWithRedux = (
  component,
  { initialState = {}, store = createStore(movies, initialState) } = {},
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

describe('<PosterItem />', () => {
  it('matches with snapshot', () => {
    const { baseElement } = renderWithRedux(
      <ModalsContextProvider>
        <Router>
          <PosterItem
            genre={poster.genres}
            poster={poster}
            hideEdit={() => {}}
            hideDelete={() => {}}
            setLoadingMovieDetails={false}
            setErrorMovieDetails={false}
            setMovieDetails={poster}
          />
        </Router>
      </ModalsContextProvider>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('shows poster title', () => {
    const posterItemComponent = renderWithRedux(
      <ModalsContextProvider>
        <Router>
          <PosterItem
            genre={poster.genres}
            poster={poster}
            hideEdit={() => {}}
            hideDelete={() => {}}
            setLoadingMovieDetails={false}
            setErrorMovieDetails={false}
            setMovieDetails={poster}
          />
        </Router>
      </ModalsContextProvider>,
    );

    expect(() => posterItemComponent.getByText(/fifty shades freed/i));
  });
});
