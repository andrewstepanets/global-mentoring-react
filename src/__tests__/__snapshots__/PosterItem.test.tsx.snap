// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PosterItem /> matches with snapshot 1`] = `
<body>
  <div>
    <li
      class="sc-iCfMLu bNnaVt"
    >
      <div
        class="sc-bdvvtL hfOjlw"
      >
        <button
          class="sc-gKclnd haiMtS"
        >
          <span
            class="sc-hKwDye eJmhdO"
          />
          <span
            class="sc-hKwDye eJmhdO"
          />
          <span
            class="sc-hKwDye eJmhdO"
          />
        </button>
      </div>
      <a
        class="sc-furwcr enHelU"
        href="/film/337167"
      >
        <img
          alt="Fifty Shades Freed"
          class="sc-iqseJM fXdUqG"
          src="https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg"
        />
        <div
          class="sc-pVTFL cGrwuW"
        >
          <h2
            class="sc-jrQzAO jtEcVe"
          >
            Fifty Shades Freed
          </h2>
          <span
            class="sc-kDTinF cDxvgr"
          >
            2016-02-11
          </span>
        </div>
        <p
          class="sc-crHmcD foEfLe"
        >
          Animation
          Adventure
          Family
          Comedy
        </p>
      </a>
    </li>
  </div>
</body>
`;
